### ISSUE TITLE
Set a reasonable batch size in the `encode` method

### ISSUE THREAD

**Issue by BBC-Esq** [UKPLab/sentence-transformers#2551]
See my analysis here...

https://github.com/michaelfeil/infinity/discussions/160#discussioncomment-8856128

Especially with the larger models, unless a programmer specifies the ```encode kwargs``` - e.g. using ```HuggingFaceEmbeddings``` a default of 32 leads to huge inefficiencies.

**Comment by tomaarsen** [comment]
Hello!

I assumed that you were actually going to advocate for a higher default batch size. But it seems that you are not:
>  you should never use a batch size larger than 1.

Which is odd, as this clashes with your own figures, e.g. this one, which shows that increasing the batch size heavily reduces the average time per text.
![314596247-af411837-1bb9-4eca-9463-2fb17466875b](https://github.com/UKPLab/sentence-transformers/assets/37621491/d8d12300-3dff-4d09-9066-4b09334f7966)

Beyond that, it is to be expected that increasing the batch size will increase the VRAM usage. There is a clear linear relation between more data on the GPU at once (i.e., higher batch size) and higher VRAM usage.

---

I ran my own experiment quickly, and had these results:
```python
import time
from sentence_transformers import SentenceTransformer
from datasets import load_dataset

model = SentenceTransformer("all-mpnet-base-v2")
# ~5k sentences, each very short
sentences = load_dataset("mteb/stsbenchmark-sts", split="train")["sentence1"]

for batch_size in [1, 2, 4, 8, 16, 32, 64, 128, 256, 512]:
    start_t = time.time()
    model.encode(sentences, batch_size=batch_size)
    duration = time.time() - start_t
    print(f"Batch size: {batch_size}, Duration: {duration:.2f} seconds; {len(sentences) / duration:.2f} sentences per second")
```
```
Batch size: 1, Duration: 74.24 seconds; 77.44 sentences per second
Batch size: 2, Duration: 41.68 seconds; 137.93 sentences per second
Batch size: 4, Duration: 19.51 seconds; 294.71 sentences per second
Batch size: 8, Duration: 10.09 seconds; 569.93 sentences per second
Batch size: 16, Duration: 5.32 seconds; 1080.97 sentences per second
Batch size: 32, Duration: 3.42 seconds; 1683.33 sentences per second
Batch size: 64, Duration: 2.39 seconds; 2409.71 sentences per second
Batch size: 128, Duration: 2.01 seconds; 2865.35 sentences per second
Batch size: 256, Duration: 1.81 seconds; 3175.07 sentences per second
Batch size: 512, Duration: 1.86 seconds; 3094.70 sentences per second
```
As expected, on my RTX 3090 I get a lot more performance when increasing the batch size. As for `HuggingFaceEmbeddings`, it is a shame that the encode kwargs do not clearly list the valid options.

In conclusion, I don't see any motivation to lower the batch size, nor any reason to increase it (a default batch size that is too high may cause troubles for people with smaller GPUs or with CPUs, which perform worse at very high batch sizes). I also won't update it as it may break backwards compatibility for users (e.g. they suddenly get OutOfMemory exceptions).

Additionally, I would advise taking a different tone in GitHub issues, discussions & pull requests.

- Tom Aarsen

**Comment by BBC-Esq** [comment]
Thanks for the response.  It was just a semi-urgent issue, and it still is IMHO, but feel free to change the title if you think it's more PC.  Thanks for copying "cc" MIchaelFeil, as if he's my Supervisor at work or something who's responsible for my posts and the "tone" of them?  lol.

Anyhow, in all seriousness, I'm not sure how you are getting such divergent results.  Is the above script that you pasted the entire script from start to finish?  It's always possible that I made a mistake in my testing rubric and I'm open to being wrong...Here's my entire script that I used so maybe you an point out any errors I made.  Be gentle though...I'm not a programmer by trade.  ;-)

Ultimately, my goal is not to be "right" but rather to give accurate results so...any corrections would be much appreciated!

Regarding your statement "Which is odd, as this clashes with your own figures, e.g. this one, which shows that increasing the batch size heavily reduces the average time per text."  What my testing showed is that as you increase it beyond the 4 or 8, those lines start going upwards - i.e. more time.  So that chart doesn't show it...I just tested the batch size up to where it flatlines.  But again, you got different results and I'd like to get to the bottom of this.

Here's my entire script sir for your review.  Be gentle, I'm not a programmer by trade.  ;-)  BTW, I love the ```sentence-transformers``` library:

<details>
<summary>CLICK TO EXPAND SCRIPT</summary>

```
import os
import shutil
from pathlib import Path
from langchain_community.embeddings import HuggingFaceEmbeddings
from langchain.text_splitter import RecursiveCharacterTextSplitter
from sentence_transformers import SentenceTransformer, util
import numpy as np
import torch
import time
import pynvml
import gc

def cleanup(model_instance=None):
    if model_instance is not None:
        del model_instance
    gc.collect()
    if torch.cuda.is_available():
        torch.cuda.empty_cache()

class CustomInstructor(SentenceTransformer):
    def __init__(self, model_name_or_path="fdaf", device='cuda'): # , dtype=torch.float16
        super().__init__(model_name_or_path, device=device)
        #if torch.cuda.is_available() and device == 'cuda' and dtype == torch.float16:
            #self.half()
        
    def encode(self, sentences, batch_size=32, show_progress_bar=False, convert_to_numpy=True, normalize_embeddings=True, **kwargs):
        self.eval()
        
        pynvml.nvmlInit()
        handle = pynvml.nvmlDeviceGetHandleByIndex(0)
        max_vram_usage = 0
        last_check_time = time.time()

        all_embeddings = []
        for start_index in range(0, len(sentences), batch_size):
            sentences_batch = sentences[start_index:start_index+batch_size]
            features = self.tokenize(sentences_batch)
            features = util.batch_to_device(features, self.device)
            
            with torch.no_grad():
                out_features = self.forward(features)
                embeddings = out_features['sentence_embedding']
                
                if normalize_embeddings:
                    embeddings = torch.nn.functional.normalize(embeddings, p=2, dim=1)
                
                all_embeddings.append(embeddings)

            current_time = time.time()
            if current_time - last_check_time >= 0.5:
                memory_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
                current_vram_usage = round(memory_info.used / (1024**3), 2)
                max_vram_usage = max(max_vram_usage, current_vram_usage)
                last_check_time = current_time
        
        all_embeddings = torch.cat(all_embeddings, 0)
        
        if convert_to_numpy:
            all_embeddings = all_embeddings.cpu().numpy()

        memory_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
        current_vram_usage = round(memory_info.used / (1024**3), 2)
        max_vram_usage = max(max_vram_usage, current_vram_usage)

        return all_embeddings, max_vram_usage

def extract_text(file_path):
    with open(file_path, 'r', encoding="utf-8") as file:
        text = file.read()
    
    return text

def split_text(text):
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=800,
        chunk_overlap=200,
    )

    text_chunks = text_splitter.split_text(text)
    
    return text_chunks

def test_batch_sizes(upper_limit, text_chunks, model_name):
    pynvml.nvmlInit()
    handle = pynvml.nvmlDeviceGetHandleByIndex(0)
    
    for batch_size in range(1, upper_limit + 1):
        vram_usages = []
        times = []
        for _ in range(3):
            model = CustomInstructor(model_name, device='cuda') #, dtype=torch.float16
            start_time = time.time()
            
            _, max_vram_usage = model.encode(text_chunks, batch_size=batch_size, normalize_embeddings=True)
            elapsed_time = time.time() - start_time
            
            vram_usages.append(max_vram_usage)
            times.append(elapsed_time)
            
            cleanup(model)
            
            time.sleep(3)

        avg_vram_usage = sum(vram_usages) / len(vram_usages)
        avg_time = sum(times) / len(times)
        print(f"Batch Size: {batch_size}, Average Max VRAM Usage: {avg_vram_usage:.2f} GB, Average Time: {avg_time:.2f} s")

if __name__ == "__main__":
    file_path = r"[path to .txt file to process]"
    text = extract_text(file_path)
    text_chunks = split_text(text)
    model_name = r"[path to model on computer]"
    upper_limit = 12
    
    test_batch_sizes(upper_limit, text_chunks, model_name)
```
</details>

**Comment by tomaarsen** [comment]
> Thanks for copying "cc" MIchaelFeil, as if he's my Supervisor at work or something who's responsible for my posts and the "tone" of them?

The cc was to involve Michael in the discussion on batch sizes, separate from the sentence above it. He has (via his https://github.com/michaelfeil/infinity project) shown to be knowledgeable on this topic.

Thanks for providing your script! It should be valuable to determine why our results differ. For example, if my code is the outlier & most people experience your findings, then changing the batch size might indeed be beneficial. I'll run your code when I have a bit of time spare & I'll get back to you.

- Tom Aarsen

**Comment by BBC-Esq** [comment]
Thanks for the feedback guys.  I'm looking forward to hearing Tom's results from my script.  There are a lot of variables at play for sure...First, there are a lot of uses for ```sentence-transformers```.  To encode multiple single sentences, sure you can have an astronomical batch size like 512.  However, in my testing I made it clear that the chunk was split into ~800 characters using the RecursiveCharacterTextSplitter, which as a rule of thumb of 4characters/token, equals 200 tokens.  Thus, 64 is completely unreasonable for RAG applications.  True, 800 is a little on the high end, but even with 600 characters that's still 150 tokens (not 64).

Second, look at the models coming out.  They have context windows in the thousands now.  Having a batch size of 512 with each chunk being, let's say, 1,000 tokens, seems to inevitably lead to inefficiency and unnecessarily increased memory usage.  I'm talking about the default value - i.e. for users who might not know how to change it...like I was.

Third, I only created charts where the efficiencies flatlined when encodding, again, approximately 200 tokens in a "chunk..." or "batch."  I tested a little bit beyond this to make sure I wasn't off base, and yes, every single model INCREASED processing time and INCREASED vram usage after you went beyond the optimal batch size.

Here's how I currently set things in my RAG application based off this research:

```
    def initialize_vector_model(self, embedding_model_name, config_data):
        compute_device = config_data['Compute_Device']['database_creation']
        model_kwargs = {"device": compute_device}
        encode_kwargs = {'normalize_embeddings': False, 'batch_size': 8}

        if compute_device.lower() == 'cpu':
            encode_kwargs['batch_size'] = 2
        else:
            batch_size_mapping = {
                'instructor-xl': 2,
                'instructor-large': 3,
                ('jina-embedding-l', 'bge-large', 'gte-large', 'roberta-large'): 4,
                'jina-embedding-s': 9,
                ('bge-small', 'gte-small'): 10,
                ('MiniLM',): 20,
            }
```

Doing it this way keeps my CUDA usage at approximately 85% on an RTX 4090, which leaves room for users with lower-end GPUs to use my program without inevitably leading to inefficiencies and increased memory usage...

Do please let me know if I'm messed up in my testing somehow.  Thanks again.

**Comment by BBC-Esq** [comment]
Can you clarify what you mean by this? "I am sorry, but you are wrong here. Making the batch_size not a multiple of 8 or power of 2 is making your benchmark untrustworthy."

Also, the tokens processed are all the same.  Each test the same text was split into the same exact chunk sizes, all within the sequence_length of the model.  I'd love for you to actually use my script or explain why my actual script is not doing what I think it is...

But if there's something I missed as far as multiples of 8, I'm not aware of that concept yet...

**Comment by BBC-Esq** [comment]
Either of you please test my script for yourselves and let me know if you get different results.  This theoretical conceptual stuff is good and all but...Unless there's something incorrect with my testing procedure and/or script, my results are reliable.  I have the data.

**Comment by BBC-Esq** [comment]
Here are the characteristics of my text that's chunked so you can come close to to replicating it...Contact me directly NOT ON HERE if you want a copy of the actual .txt file.  I won't give it here for copyright reasons, but legally believe it's OK for research purposes like we're doing.

But here are the general characteristics.  Make sure an use a comparable .txt file of your own for comparable chunks sizes and results:

* Legal treatise with various chapters and sections...we're not dealing with an EXCEL spreadsheet here.
* 2,598,932 characters on 57,001 lines
* Remember, chunks size of 800 using the RecursivCharacterTextSplitter
* File itself is .txt and the code to extract and chunk I've already given.

Again, to ensure apples-to-apples, contact me directly to avoid copyright concerns.

**Comment by BBC-Esq** [comment]
I'd also like to know what the size of the sentences are in Tom's script since he's claiming speed improvements up to 512 even...And Tom, can did you keep track of VRAM usage as well?

```
import time
from sentence_transformers import SentenceTransformer
from datasets import load_dataset

model = SentenceTransformer("all-mpnet-base-v2")
# ~5k sentences, each very short
sentences = load_dataset("mteb/stsbenchmark-sts", split="train")["sentence1"]

for batch_size in [1, 2, 4, 8, 16, 32, 64, 128, 256, 512]:
    start_t = time.time()
    model.encode(sentences, batch_size=batch_size)
    duration = time.time() - start_t
    print(f"Batch size: {batch_size}, Duration: {duration:.2f} seconds; {len(sentences) / duration:.2f} sentences per second")
```

**Comment by tomaarsen** [comment]
Sure, here it is:
```python
import time
import torch
from sentence_transformers import SentenceTransformer
from datasets import load_dataset

model = SentenceTransformer("all-mpnet-base-v2")
sentences = load_dataset("mteb/stsbenchmark-sts", split="train")["sentence1"]

for batch_size in [1, 2, 4, 8, 16, 32, 64, 128, 256, 512]:
    start_t = time.time()
    model.encode(sentences, batch_size=batch_size)
    duration = time.time() - start_t
    print(f"Batch size: {batch_size}, Duration: {duration:.2f} seconds; {len(sentences) / duration:.2f} sentences per second; {torch.cuda.memory_allocated() / 1024**3:.3f} GB VRAM usage")
```
```
Batch size: 1, Duration: 74.45 seconds; 77.22 sentences per second; 0.419GB max VRAM usage
Batch size: 2, Duration: 42.31 seconds; 135.87 sentences per second; 0.422GB max VRAM usage
Batch size: 4, Duration: 22.04 seconds; 260.79 sentences per second; 0.427GB max VRAM usage
Batch size: 8, Duration: 11.80 seconds; 487.08 sentences per second; 0.437GB max VRAM usage
Batch size: 16, Duration: 6.03 seconds; 953.30 sentences per second; 0.458GB max VRAM usage
Batch size: 32, Duration: 3.23 seconds; 1779.35 sentences per second; 0.495GB max VRAM usage
Batch size: 64, Duration: 2.33 seconds; 2462.62 sentences per second; 0.574GB max VRAM usage
Batch size: 128, Duration: 2.02 seconds; 2852.20 sentences per second; 0.729GB max VRAM usage
Batch size: 256, Duration: 1.81 seconds; 3176.26 sentences per second; 1.038GB max VRAM usage
Batch size: 512, Duration: 1.87 seconds; 3072.13 sentences per second; 1.660GB max VRAM usage
```
As expected, higher batch size => more data on the GPU => higher VRAM usage.

As for your script, these are my results:
```
Batch Size: 1, Average Max VRAM Usage: 2.79 GB, Average Time: 0.15 s
Batch Size: 2, Average Max VRAM Usage: 2.93 GB, Average Time: 0.08 s
Batch Size: 3, Average Max VRAM Usage: 2.92 GB, Average Time: 0.08 s
Batch Size: 4, Average Max VRAM Usage: 2.91 GB, Average Time: 0.09 s
Batch Size: 5, Average Max VRAM Usage: 2.92 GB, Average Time: 0.04 s
Batch Size: 6, Average Max VRAM Usage: 2.91 GB, Average Time: 0.09 s
Batch Size: 7, Average Max VRAM Usage: 2.91 GB, Average Time: 0.08 s
Batch Size: 8, Average Max VRAM Usage: 2.92 GB, Average Time: 0.07 s
Batch Size: 9, Average Max VRAM Usage: 2.94 GB, Average Time: 0.08 s
Batch Size: 10, Average Max VRAM Usage: 2.90 GB, Average Time: 0.04 s
Batch Size: 11, Average Max VRAM Usage: 2.90 GB, Average Time: 0.11 s
Batch Size: 12, Average Max VRAM Usage: 2.91 GB, Average Time: 0.10 s
```

This seems to indicate that it makes more sense to have a higher batch size. Also, I think your VRAM measurement is a bit off: you're measuring the "current" memory usage. If I do that in my script (with `torch.cuda.memory_allocated` instead of `torch.cuda.max_memory_allocated`), then I'll get 0.471GB across the board, because the VRAM used for the batches has already been cleaned up again.
Also, the difference in VRAM (0.471GB vs 2.90GB) can be attributed to the overhead by PyTorch/CUDA being measured in your setup, but ignored in mine.
> Making the batch_size not a multiple of 8 or power of 2 is making your benchmark untrustworthy."

Multiples of 8 play much better in GPUs as they can be more easily split across the GPU.

I will be following @michaelfeil his advice and keep the batch size at 32. 

- Tom Aarsen

**Comment by BBC-Esq** [comment]
I'm not using ```torch.cuda.memory_allocated``` nor ```torch.cuda.max_memory_allocated```, but rather Nvidia's library to interact with ```nvml```.  It polls the total vram usage every .5 seconds while the vector model is encoding and then takes the highest point of all polls and returns that value, and it does this for each run and then creates an average.

It's my understanding that if you use ```torch```, however, it does not return overall vram usage in general, just the vram being used for pytorch operations.  I was clear in my testing notes that my numbers took into account this overhead - i.e. other things on my system using vram.  The import thing is that the vram usage is still accurate "relative" to one another (assuming there's not a significant change in my "overhead").  I used this approach to simulate a more real world testing environment, how a user would use the model.

Do you still think that my vram numbers are off considering that you and I apparently didn't employ the same measurement metric apparently?

Regarding your usage of my script, you stated that you received these results:

```
Batch Size: 9, Average Max VRAM Usage: 2.94 GB, Average Time: 0.08 s
Batch Size: 10, Average Max VRAM Usage: 2.90 GB, Average Time: 0.04 s
Batch Size: 11, Average Max VRAM Usage: 2.90 GB, Average Time: 0.11 s
Batch Size: 12, Average Max VRAM Usage: 2.91 GB, Average Time: 0.10 
```

How does this NOT confirm my findings?  At the optimal batch size of 8-10, the compute time actually starts increasing just like in my testing.  Thus, it stands to reason that setting a ```batch_size``` of 32 would increase it even further.  Also, which vector model were you testing, all-mpnet-base-v2?  Using an RTX 4090?  Windows or Linux?

Apart from the compute time increasing after the optimal batch size, however, I'm mystified how your vram usage did not actually start increasing.  Maybe there's a difference in gpu used or other testing setup software/hardware?

@michaelfeil Did you actually do any testing?

I did some additional testing with ```sentect-t5-xl``` yesterday, and anything beyond a batch size of 1 makes it 3x as slow and anything beyond that borks the entire process...If I'm mistaken in my implementation of ```sentence-transformers``` and/or testing I'd like a little more explanation.  Your testing partially confirms my results and doesn't include the necessary details for me to evaluate whether there's been a true apples-to-apples testing setup (or as close as feasible...).

**Comment by tomaarsen** [comment]
> Also, which vector model were you testing, all-mpnet-base-v2? Using an RTX 4090? Windows or Linux?
`all-mpnet-base-v2`, RTX 3090, Windows. 

> Thus, it stands to reason that setting a batch_size of 32 would increase it even further.

Not in practice, it seems:
```
Batch Size: 32, Average Max VRAM Usage: 2.76 GB, Average Time: 0.11 s
Batch Size: 64, Average Max VRAM Usage: 2.72 GB, Average Time: 0.12 s
Batch Size: 128, Average Max VRAM Usage: 2.73 GB, Average Time: 0.12 s
Batch Size: 256, Average Max VRAM Usage: 2.76 GB, Average Time: 0.10 s
Batch Size: 512, Average Max VRAM Usage: 2.74 GB, Average Time: 0.07 s
```

I took a bunch more data this time around:
```
Batch Size: 1, Average Max VRAM Usage: 2.73 GB, Average Time: 8.46 s
Batch Size: 2, Average Max VRAM Usage: 2.74 GB, Average Time: 4.77 s
Batch Size: 4, Average Max VRAM Usage: 2.73 GB, Average Time: 2.80 s
Batch Size: 8, Average Max VRAM Usage: 2.76 GB, Average Time: 2.21 s
Batch Size: 16, Average Max VRAM Usage: 2.84 GB, Average Time: 1.83 s
Batch Size: 32, Average Max VRAM Usage: 3.20 GB, Average Time: 1.65 s
Batch Size: 64, Average Max VRAM Usage: 4.04 GB, Average Time: 1.64 s
Batch Size: 128, Average Max VRAM Usage: 5.36 GB, Average Time: 1.64 s
Batch Size: 256, Average Max VRAM Usage: 6.44 GB, Average Time: 1.67 s
Batch Size: 512, Average Max VRAM Usage: 10.09 GB, Average Time: 1.88 s
```

Also, my apologies, your VRAM measurements look right - I just didn't use enough data the first time around to notice the difference.

**Comment by tomaarsen** [comment]
> you did in-fact run my script verbatim...no changes?

Yes, indeed. Only the data is different, of course, I had 564 text chunks in total I believe.

> Also, can you explain briefly why multiples of 8 might be more efficient?

In essence, GPUs are a massive collection of tiny "CPUs". Despite being slow individually, these tiny "CPUs" are extremely parallelizable, which is why GPUs are fast for those tasks. There are essentially always a multiple of 8 of these "CPUs", and so most GPU processing is done in batches of multiples of 8, as then the "tasks" can be scheduled across the GPU components more evenly and efficiently. Apparently that isn't as much of an issue nowadays as GPUs have improved, but the custom of "multiple of 8" or "power of 2" has stayed for the most part.

This is my variation of your script with my data:

```python
import os
import shutil
from pathlib import Path
from langchain_community.embeddings import HuggingFaceEmbeddings
from langchain.text_splitter import RecursiveCharacterTextSplitter
from sentence_transformers import SentenceTransformer, util
import numpy as np
import torch
import time
import pynvml
import gc
from datasets import load_dataset

def cleanup(model_instance=None):
    if model_instance is not None:
        del model_instance
    gc.collect()
    if torch.cuda.is_available():
        torch.cuda.empty_cache()

class CustomInstructor(SentenceTransformer):
    def __init__(self, model_name_or_path="fdaf", device='cuda'): # , dtype=torch.float16
        super().__init__(model_name_or_path, device=device)
        #if torch.cuda.is_available() and device == 'cuda' and dtype == torch.float16:
            #self.half()
        
    def encode(self, sentences, batch_size=32, show_progress_bar=False, convert_to_numpy=True, normalize_embeddings=True, **kwargs):
        self.eval()
        
        pynvml.nvmlInit()
        handle = pynvml.nvmlDeviceGetHandleByIndex(0)
        max_vram_usage = 0
        last_check_time = time.time()

        all_embeddings = []
        for start_index in range(0, len(sentences), batch_size):
            sentences_batch = sentences[start_index:start_index+batch_size]
            features = self.tokenize(sentences_batch)
            features = util.batch_to_device(features, self.device)
            
            with torch.no_grad():
                out_features = self.forward(features)
                embeddings = out_features['sentence_embedding']
                
                if normalize_embeddings:
                    embeddings = torch.nn.functional.normalize(embeddings, p=2, dim=1)
                
                all_embeddings.append(embeddings)

            current_time = time.time()
            if current_time - last_check_time >= 0.5:
                memory_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
                current_vram_usage = round(memory_info.used / (1024**3), 2)
                max_vram_usage = max(max_vram_usage, current_vram_usage)
                last_check_time = current_time
        
        all_embeddings = torch.cat(all_embeddings, 0)
        
        if convert_to_numpy:
            all_embeddings = all_embeddings.cpu().numpy()

        memory_info = pynvml.nvmlDeviceGetMemoryInfo(handle)
        current_vram_usage = round(memory_info.used / (1024**3), 2)
        max_vram_usage = max(max_vram_usage, current_vram_usage)

        return all_embeddings, max_vram_usage

def extract_text(file_path):
    with open(file_path, 'r', encoding="utf-8") as file:
        text = file.read()
    
    return text

def split_text(text):
    text_splitter = RecursiveCharacterTextSplitter(
        chunk_size=800,
        chunk_overlap=200,
    )

    text_chunks = text_splitter.split_text(text)
    
    return text_chunks

def test_batch_sizes(upper_limit, text_chunks, model_name):
    pynvml.nvmlInit()
    handle = pynvml.nvmlDeviceGetHandleByIndex(0)
    
    for batch_size in [upper_limit]:
        vram_usages = []
        times = []
        for _ in range(3):
            model = CustomInstructor(model_name, device='cuda') #, dtype=torch.float16
            start_time = time.time()
            _, max_vram_usage = model.encode(text_chunks, batch_size=batch_size, normalize_embeddings=True)
            elapsed_time = time.time() - start_time
            
            vram_usages.append(max_vram_usage)
            times.append(elapsed_time)
            
            cleanup(model)
            
            # time.sleep(3)

        avg_vram_usage = sum(vram_usages) / len(vram_usages)
        avg_time = sum(times) / len(times)
        print(f"Batch Size: {batch_size}, Average Max VRAM Usage: {avg_vram_usage:.2f} GB, Average Time: {avg_time:.2f} s")

if __name__ == "__main__":
    sentences = load_dataset("mteb/stsbenchmark-sts", split="train")["sentence1"]
    text = "\n".join(sentences[:10000])
    text_chunks = split_text(text)
    model_name = r"all-mpnet-base-v2"
    upper_limit = 16

    test_batch_sizes(upper_limit, text_chunks, model_name)
```

Also, you mention performance on specific hardware, with specific kinds of text, with specific text topics, with specific model sizes, but realise that the default is intended to work as a default for the widest amount of users in the widest amount of circumstances. That's also why I wouldn't change the batch size because on one specific GPU, one specific model and one specific text length, another batch size works better. E.g. according to https://github.com/UKPLab/sentence-transformers/issues/2551#issuecomment-2020124671, a batch size of 256 is actually better for that setup. I normally use `batch_size=512` or `256` when quickly embedding some text for a test.

- Tom Aarsen

**Comment by BBC-Esq** [comment]
1) Thanks for the quick explanation about 8/2power...will read more into that;

2) Totally understand about setting the default where "most" usages will benefit...I see that as having already changed with the newer models and usages, but that's a subject argument so...And that's why I qualified my testing as specific for 
RAG purposes.  I'm not as familiar with the other uses cases or if they recommend much lower chunk sizes...and like I said, even going from 800-600 there was significant reduction in compute requirement for me...

3) Can you please do one final test some kind of text that has approximately 2,598,932 "characters" before it's split into chunks, like I did?  I plan on retesting significantly and this would save me some time in eliminating this variable...I processed thousands of chunks...

In conclusion, I initially felt that you closed out this issue prematurely and were making short shrift of my dedicated time spent on testing that care about this issue in general.  However, the fact that you responded as well as reading the substance of those responses, no leaves me with the impression that you truly do care about this issue and have put a lot of good thought into it.  I enjoy testing these things and learning about this technology in general, as I know you do, and there's always more for me to learn so...
